src/components/transition/CSSTransitionAnimator.js

```diff
@@ -0,0 +1,205 @@
+        transform: hasChangedValue('transform'),
+        width: hasChangedValue('width'),
+        height: hasChangedValue('height'),
+        opacity: hasChangedValue('opacity'),
+      });
+    }
+  }
+
+  function animate(
+    element: HTMLElement,
+    from?: PropertyObjectType,
+    to?: PropertyObjectType,
+    speed?: number
+  ) {
+    if (from !== undefined) {
+      pushHistoryState(parsePropertyObject(from));
+    }
+    if (to !== undefined) {
+      pushHistoryState(parsePropertyObject(to));
+    }
+
+    const willChangeProps = {
+      transform: hasChangedValue('transform'),
+      width: hasChangedValue('width'),
+      height: hasChangedValue('height'),
+      opacity: hasChangedValue('opacity'),
+    };
+
+    remainingPropsToChange = Object.keys(willChangeProps).reduce(
+      (sum, prop) => sum + Number(willChangeProps[prop]),
+      0
+    );
+
+    if (from !== undefined) {
+      addElementStyles(element, parsedPropsHistory[0], willChangeProps, speed);
+    }
+    if (to !== undefined) {
+      addElementStyles(element, parsedPropsHistory[1], willChangeProps, speed);
+    }
+  }
+
+  function addElementStyles(
+    element: HTMLElement,
+    parsedProps: ParsedPropertyObjectType,
+    willChangeProps: WillChangePropsType,
+    speed?: number
+  ) {
+    const {className, transform, width, height, opacity} = parsedProps;
+    const transitionProperty = [];
+    const transitionDuration = [];
+    const transitionTimingFunction = [];
+
+    /**
+     * The transition values should follow the order
+     * of properties in each CSS Transition array, which
+     * is ensured by arrays in a loop.
+     */
+    Object.keys(willChangeProps).forEach(prop => {
+      if (willChangeProps[prop]) {
+        transitionProperty.push(prop);
+        transitionDuration.push(applySpeed(parsedProps[prop].duration, speed));
+        transitionTimingFunction.push(parsedProps[prop].easing);
+      }
+    });
+
+    const combine = (a: Array<string>) => a.join(', ');
+
+    classNamesRegistry.register('transition', className);
+    element.className = classNamesRegistry.toString();
+    element.style.willChange = combine(transitionProperty);
+    element.style.transitionProperty = combine(transitionProperty);
+    element.style.transitionDuration = combine(oneOrAll(transitionDuration));
+    element.style.transitionTimingFunction = combine(
+      oneOrAll(transitionTimingFunction)
+    );
+
+    if (willChangeProps.transform) {
+      element.style.transform = transform.value;
+      element.style.transformOrigin = transform.origin;
+    }
+
+    if (willChangeProps.width) {
+      element.style.width = width.value === 'auto' ? '' : width.value;
+    }
+
+    if (willChangeProps.height) {
+      element.style.height = height.value === 'auto' ? '' : height.value;
+    }
+
+    if (willChangeProps.opacity) {
+      element.style.opacity = opacity.value;
+    }
+
+    // repaint in case of further changes
+    forceRepaint(element);
```

Comment:
I am not sure why we need to sync reflow here. Comment above is not really helpful

Reply:
It should be called between sync `addElementStyles()`. I can move it where they are called for better understanding.

Reply:
Thx

Reply:
@coderitual btw. I just found a bug related to this part, but the fix should be soon (this week).